### PR TITLE
[ADD] udes_13: Add get_info submodule

### PR DESCRIPTION
**Pull request by peterpratt13**
Have a seperate module that determines how information can be recursively obtianed.

User-story: 9944

Signed-off-by: Pete Pratt <peter.pratt@unipart.io>

### CODE REVIEW

File: addons/get_info/models/models.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+from addons.udes_common.models import add_if_not_exists
+
+
+# N.B: Basic values should not have any relations as this is to stop the recursion
+BASIC_GET_INFO_VALUES = {"id", "name", "display_name"}
+
+
+@add_if_not_exists(models.BaseModel)
+def _get_info(self, level=0, max_level=None, fields=None):
+
```
Comment (peterpratt13) [inline addons/get_info/models/models.py:12]: Add doc string

File: addons/get_info/tests/test_get_info.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    def test02_check_keys(self):
+        """ Check we get the expected keys from the _get_info_field_names field """
+        self.assertEqual(
+            self.grandchild_location.get_info()[0].keys(), self.Location._get_info_field_names
+        )
+
+    def test03_get_info_with_additional_levels_simple(self):
+        """ Check the recursion levels by adding field to default fields
+        """
```
Comment (peterpratt13) [inline addons/get_info/tests/test_get_info.py:85]: Fix

File: addons/get_info/tests/test_get_info.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        self._check_fields_helper(test_fields, info)
+
+    def test05_get_info_with_different_recursion(self):
+        """ Check the recursion levels by adding field to default fields """
+        # Define fields to search for simplicity
+        test_fields = {"location_id", "name"}
+        # Run checks via helper
+        info_level1 = self.grandchild_location.get_info(max_level=1, fields=test_fields)[0]
+        info_level3 = self.grandchild_location.get_info(max_level=3, fields=test_fields)[0]
+        # Check no entries with level 1 has a dictionary
```
Comment (peterpratt13) [inline addons/get_info/tests/test_get_info.py:110]: Add additional comment for bit below, or rephrase

File: addons/get_info/tests/test_get_info.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        self._recursive_field_helper("location_id", info, self.family_tree)
+
+    def test04_get_info_with_custom_fields(self):
+        """ Check the recursion levels by adding field to default fields
+        """
+        # Define fields to search
+        test_fields = {"location_id", "name"}
+        # Run checks via helper
+        info = self.grandchild_location.get_info(max_level=3, fields=test_fields)[0]
+        # # Check
```
Comment (peterpratt13) [inline addons/get_info/tests/test_get_info.py:99]: Double #

File: addons/get_info/models/models.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    for field_name in fields:
+        # Allow for name to be changed in info output
+        if isinstance(field_name, tuple):
+            info_name, field_name = field_name
+        else:
+            info_name = field_name
+
+        if field_name not in self:
+            # skip over fields which don't exist
+            # perhaps raise a warning here
```
Comment (ws-mps) [inline addons/get_info/models/models.py:31]: Yes please, raise a warning at the logs.

File: addons/get_info/tests/test_get_info.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+@common.at_install(False)
+@common.post_install(True)
+class TestStockLocation(common.SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super(TestStockLocation, cls).setUpClass()
+        cls.Location = cls.env["stock.location"]
+        # Get stock location
+        cls.stock_location = cls.Location.search([]).filtered(lambda l: l.name == "Stock")
```
Comment (ws-mps) [inline addons/get_info/tests/test_get_info.py:16]: Why don't we use cls.env['XXX']?

Comment (peterpratt13) [reply addons/get_info/tests/test_get_info.py:16]: Using the cls.Location above. But only use Location once elsewhere, so tidied this up

**Pull request merged**